### ISSUE TITLE
Wrong backward algorithm implementation?

### ISSUE THREAD

**Issue by viktor-shcherb** [kajyuuen/pytorch-partial-crf#21]
I might be wrong but shouldn't forward algorithm go from `1` to `seq_len - 1` with already computed element in `log_proba` being first emission + start transition? And shouldn't backward algorithm go from `seq_len - 2` to `0` with already computed element in `log_proba` being the last emission + end transition?

As of now your backward algorithm goes from `seq_len - 1` to `1` with already computed element being just the end transition.

I apologize if I misunderstood your code or the formulas [here](https://towardsai.net/p/machine-learning/a-tour-of-conditional-random-field-7d8476ce0201).

**Comment by gcgbarbosa** [comment]
Would you mind pointing out to the lines of code you think should be changed?

**Comment by kajyuuen** [comment]
@ViktorooReps -san, @gcgbarbosa -san, Thanks for your comment.
The test about marginal probabilities using forward and backward algorithms to produce has passed.

https://github.com/kajyuuen/pytorch-partial-crf/blob/bb368346544cb6241e425a8b4e1a3baee324ef0d/tests/test_base_crf.py#L95-L98

But there may be potential errors.
Could you send me some specific examples here much or send me a PR to add a test cases?

**Comment by viktor-shcherb** [comment]
Here is the code that should theoretically work, but it fails:

        def test_marginal_probabilities(self):
                marginal_probabilities = self.base_crf.marginal_probabilities(self.emissions)
                # TODO: Add test
                assert torch.allclose(marginal_probabilities.sum(dim=2), torch.ones_like(marginal_probabilities.sum(dim=2)))
        
                predicted_path = torch.argmax(marginal_probabilities.transpose(0, 1), dim=-1)
                viterbi_path = torch.tensor(self.base_crf.viterbi_decode(self.emissions))
                assert torch.allclose(predicted_path, viterbi_path)
        
                random_emissions = torch.rand((30, 30, 6))
                random_path = torch.tensor(self.base_crf.viterbi_decode(random_emissions))
                predicted_path = torch.argmax(self.base_crf.marginal_probabilities(random_emissions).transpose(0, 1), dim=-1)
                assert (random_path == predicted_path).all().item()
                
I've tried to make it work but without any luck for now.

**Comment by viktor-shcherb** [comment]
> Would you mind pointing out to the lines of code you think should be changed?

I've changed these lines and after that everything stopped working whatsoever. It seems like I'm misunderstanding something. Still the aformentioned code should work, so there should be a small mistake in probabilites computation.

**Comment by viktor-shcherb** [comment]
Also after initializing crf parameters randomly, predictions of `viterby_decode` do not match predictions made with `marginal_probabilities`.

**Comment by kajyuuen** [comment]
@ViktorooReps -san
Thank you for the detailed report.

there is certainly some error in the calculation of `marginal_probabilities`, I guess.
I would like to fix it, but I may be currently too busy to fix it right away.
If you don't mind, could you please correct this in PR?

Also, why don't you take this opportunity to join this repository as a maintainer?

**Comment by gcgbarbosa** [comment]
I can help @ViktorooReps with testing and stuff if he hops in.

**Comment by viktor-shcherb** [comment]
After some research I figured out that path from argmax of marginal probabilities is not the same as path from Viterbi algorithm. Basically output of Viterbi is the most probable path while marginal probability is the general probability of tag appearing on that position in all possible sequences of tags.

Keras even has different loss implementations for crf - one through negative log likelihood (which is optimal in terms of statistics) and one through categorical cross-entropy loss (using marginal probabilities).

They even go so far as stating that it is not recommended to obtain predictions with marginal probabilities while training with negative log likelihood (Viterbi is preferred for such case). 

https://github.com/keras-team/keras-contrib/blob/3fc5ef709e061416f4bc8a92ca3750c824b5d2b0/keras_contrib/layers/crf.py#L44-L53

**Comment by viktor-shcherb** [comment]
@kajyuuen, @gcgbarbosa thanks for the offer, but unfortunately I do not have time at the moment to maintain this project.

Considering this issue, I think everything is fine and my problem was that I trained my model with NLL loss and obtained predictions with marginal probabilities. Which works, but not as good as I expected. Maybe I'll attempt to train CRF with cross entropy loss for better results.

**Comment by kajyuuen** [comment]
@gcgbarbosa -san, @ViktorooReps -san

Thank you very much.
I will continue to maintain this project, but I apologize for the delay in responding as I have been busy with work recently.
I look forward to your new Issues and PRs.
If you would like to become a maintainer, please contact me.
